### PR TITLE
Sibra state processing.

### PR DESCRIPTION
**Pull request by kormat**
This change adds SIBRA state management and packet processing for the
router. The code currently only supports steady SIBRA paths. An
integration test is added to allow testing of steady path setup and
renewal.

(Please note this is a work-in-progress. Test coverage is not complete, error handling is not complete, the code is mostly uncommented)
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/netsec-ethz/scion/pull/602%23issuecomment-177884569%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23issuecomment-177939436%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23discussion_r51542744%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23discussion_r51542969%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23discussion_r51544136%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23discussion_r51550377%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23discussion_r51550799%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23discussion_r51550839%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23discussion_r51557373%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23discussion_r51560155%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23discussion_r51728921%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23discussion_r51729301%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23discussion_r51730126%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23issuecomment-179276230%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23issuecomment-179714463%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23issuecomment-177884569%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%40pszalach%20-%20a%20larger%20chunk%2C%20as%20requested%20%3A%29%22%2C%20%22created_at%22%3A%20%222016-02-01T09%3A57%3A26Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22%40reischuk%20%40perrig%20%22%2C%20%22created_at%22%3A%20%222016-02-01T12%3A02%3A24Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22LGTM%20%28of%20course%20I%20didn%27t%20read%20tests%3B-%29%22%2C%20%22created_at%22%3A%20%222016-02-03T14%3A56%3A51Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22second%20commit%20LGTM%22%2C%20%22created_at%22%3A%20%222016-02-04T08%3A55%3A18Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%207341c6d034f424188149d962434b1ec7adabc573%20lib/sibra/ext/steady.py%2058%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23discussion_r51557373%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22why%20it%20is%20padded%20here%2C%20and%20why%20by%202%20bytes%3F%5Cr%5Cn%22%2C%20%22created_at%22%3A%20%222016-02-02T11%3A40%3A50Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22https%3A//gist.github.com/kormat/49f7346dcca214b36124%23file-3-steady-txt%5Cr%5CnSteady%20reservations%20only%20have%20a%20single%20path.%20The%20other%202%20bytes%20are%20zero.%22%2C%20%22created_at%22%3A%20%222016-02-02T12%3A13%3A46Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/sibra/ext/steady.py%3AL1-202%22%7D%2C%20%22Pull%20f63d3b74ded19d0e2e71ef618a2784b0e6c0d985%20lib/sibra/state/reservation.py%2061%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23discussion_r51728921%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22What%20is%20relation%20between%20%60self.order%60%20and%20%60self.idxes%60%3F%20%60order%60%20seems%20to%20be%20just%20%60idxes.keys%28%29%60%22%2C%20%22created_at%22%3A%20%222016-02-03T14%3A45%3A53Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22%60order%60%20is%20precisely%20%60idxes.keys%28%29%60%2C%20but%20ordered.%20Dictionaries%20don%27t%20preserve%20order.%22%2C%20%22created_at%22%3A%20%222016-02-03T14%3A48%3A42Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22You%20can%27t%20simply%20sort%20the%20indexes%2C%20as%20they%20wrap.%20The%20code%20doesn%27t%20actually%20care%20about%20the%20numeric%20value%20of%20the%20indexes%20at%20all%2C%20so%20long%20as%20you%20don%27t%20try%20to%20reuse%20an%20active%20one.%22%2C%20%22created_at%22%3A%20%222016-02-03T14%3A53%3A53Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/sibra/state/reservation.py%3AL1-153%22%7D%2C%20%22Pull%20ac84c2b02b09bc22180661f52f29c73b32d03902%20lib/sibra/ext/ephemeral.py%2053%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23discussion_r51544136%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22in%20the%20future%20these%20IDs%20will%20be%20obtained%20in%20a%20special%20way%2C%20so%20maybe%20put%20a%20comment%20here%2C%20or%20create%20a%20dummy%20function%20%60_get_eph_id%28%29%60%22%2C%20%22created_at%22%3A%20%222016-02-02T09%3A38%3A04Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22They%20will%3F%22%2C%20%22created_at%22%3A%20%222016-02-02T10%3A34%3A24Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/sibra/ext/ephemeral.py%3AL1-76%22%7D%2C%20%22Pull%20ac84c2b02b09bc22180661f52f29c73b32d03902%20lib/sibra/ext/ephemeral.py%201%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23discussion_r51542969%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%222016/%3A-%29%22%2C%20%22created_at%22%3A%20%222016-02-02T09%3A29%3A22Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22Fixed%20%28in%20multiple%20places%29%22%2C%20%22created_at%22%3A%20%222016-02-02T10%3A33%3A59Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/sibra/ext/ephemeral.py%3AL1-76%22%7D%2C%20%22Pull%20ac84c2b02b09bc22180661f52f29c73b32d03902%20lib/packet/ext_util.py%205%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/602%23discussion_r51542744%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22order%20imports%22%2C%20%22created_at%22%3A%20%222016-02-02T09%3A27%3A25Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22Fixed.%22%2C%20%22created_at%22%3A%20%222016-02-02T10%3A30%3A16Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/packet/ext_util.py%3AL24-37%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/602#issuecomment-177884569'>General Comment</a></b>
- <a href='https://github.com/kormat'></a> @pszalach - a larger chunk, as requested :)
- <a href='https://github.com/kormat'></a> @reischuk @perrig
- <a href='https://github.com/pszalach'></a> LGTM (of course I didn't read tests;-)
- <a href='https://github.com/pszalach'></a> second commit LGTM
- [ ] <a href='#crh-comment-Pull ac84c2b02b09bc22180661f52f29c73b32d03902 lib/packet/ext_util.py 5'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/602#discussion_r51542744'>File: lib/packet/ext_util.py:L24-37</a></b>
- <a href='https://github.com/pszalach'></a> order imports
- <a href='https://github.com/kormat'></a> Fixed.
- [ ] <a href='#crh-comment-Pull ac84c2b02b09bc22180661f52f29c73b32d03902 lib/sibra/ext/ephemeral.py 1'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/602#discussion_r51542969'>File: lib/sibra/ext/ephemeral.py:L1-76</a></b>
- <a href='https://github.com/pszalach'></a> 2016/:-)
- <a href='https://github.com/kormat'></a> Fixed (in multiple places)
- [ ] <a href='#crh-comment-Pull ac84c2b02b09bc22180661f52f29c73b32d03902 lib/sibra/ext/ephemeral.py 53'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/602#discussion_r51544136'>File: lib/sibra/ext/ephemeral.py:L1-76</a></b>
- <a href='https://github.com/pszalach'></a> in the future these IDs will be obtained in a special way, so maybe put a comment here, or create a dummy function `_get_eph_id()`
- <a href='https://github.com/kormat'></a> They will?
- [ ] <a href='#crh-comment-Pull 7341c6d034f424188149d962434b1ec7adabc573 lib/sibra/ext/steady.py 58'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/602#discussion_r51557373'>File: lib/sibra/ext/steady.py:L1-202</a></b>
- <a href='https://github.com/pszalach'></a> why it is padded here, and why by 2 bytes?
- <a href='https://github.com/kormat'></a> https://gist.github.com/kormat/49f7346dcca214b36124#file-3-steady-txt
  Steady reservations only have a single path. The other 2 bytes are zero.
- [ ] <a href='#crh-comment-Pull f63d3b74ded19d0e2e71ef618a2784b0e6c0d985 lib/sibra/state/reservation.py 61'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/602#discussion_r51728921'>File: lib/sibra/state/reservation.py:L1-153</a></b>
- <a href='https://github.com/pszalach'></a> What is relation between `self.order` and `self.idxes`? `order` seems to be just `idxes.keys()`
- <a href='https://github.com/kormat'></a> `order` is precisely `idxes.keys()`, but ordered. Dictionaries don't preserve order.
- <a href='https://github.com/kormat'></a> You can't simply sort the indexes, as they wrap. The code doesn't actually care about the numeric value of the indexes at all, so long as you don't try to reuse an active one.

<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/602?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/602?approve=1'></a>&nbsp;<a href='https://github.com/netsec-ethz/scion/pull/602'></a>
<a href='#crh-end'></a>

### CODE REVIEW

File: lib/packet/ext_util.py [hunk 1]
```diff
@@ -24,13 +24,13 @@
 from lib.errors import SCIONParseError
 from lib.packet.ext_hdr import ExtensionHeader
 from lib.packet.ext.path_transport import PathTransportExt
-from lib.packet.ext.sibra import SibraExt
+from lib.sibra.ext.util import parse_sibra_ext
```
Comment (pszal) [inline lib/packet/ext_util.py]: order imports

File: lib/sibra/ext/ephemeral.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
+# Copyright 2015 ETH Zurich
```
Comment (pszal) [inline lib/sibra/ext/ephemeral.py]: 2016/:-)

File: lib/sibra/ext/ephemeral.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            self.active_blocks.append(self._parse_block(data, plen), self.setup)
+        self._parse_end(data)
+
+    @classmethod
+    def from_values(cls, isd_ad, req_info, steady_ids, steady_blocks):
+        assert len(steady_ids) == len(steady_blocks)
+        inst = cls()
+        inst.setup = True
+        inst.steady = False
+        inst.path_ids.append(isd_ad.pack() +
```
Comment (pszal) [inline lib/sibra/ext/ephemeral.py:65]: in the future these IDs will be obtained in a special way, so maybe put a comment here, or create a dummy function `_get_eph_id()`

File: lib/packet/ext_util.py (hunk 1 above)
Comment (kormat) [inline lib/packet/ext_util.py]: Fixed.

File: lib/sibra/ext/ephemeral.py (hunk 2 above)
Comment (kormat) [inline lib/sibra/ext/ephemeral.py]: Fixed (in multiple places)

File: lib/sibra/ext/ephemeral.py (hunk 3 above)
Comment (kormat) [inline lib/sibra/ext/ephemeral.py:65]: They will?

File: lib/sibra/ext/steady.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        inst.total_hops = total_hops
+        inst.path_ids.append(isd_ad.pack() +
+                             os.urandom(SIBRA_STEADY_ID_LEN - isd_ad.LEN))
+        inst.req_block = ResvBlockSteady.from_values(req_info, inst.total_hops)
+        inst._set_size()
+        return inst
+
+    def pack(self):
+        raw = self._pack_start()
+        raw.append(struct.pack("!Bxx", self.total_hops))
```
Comment (pszal) [inline lib/sibra/ext/steady.py:67]: why it is padded here, and why by 2 bytes?

Comment (kormat) [inline lib/sibra/ext/steady.py:67]: https://gist.github.com/kormat/49f7346dcca214b36124#file-3-steady-txt
Steady reservations only have a single path. The other 2 bytes are zero.

File: lib/sibra/state/reservation.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        assert idx in self.idxes
+        assert idx in self.order
+        del self.idxes[idx]
+        self.order.remove(idx)
+        self._update(curr_tick)
+
+    def remove_all(self, curr_tick):
+        # Used when a request is rejected
+        self.idxes = {}
+        self.order = []
```
Comment (pszal) [inline lib/sibra/state/reservation.py]: What is relation between `self.order` and `self.idxes`? `order` seems to be just `idxes.keys()`

Comment (kormat) [inline lib/sibra/state/reservation.py]: `order` is precisely `idxes.keys()`, but ordered. Dictionaries don't preserve order.

Comment (kormat) [inline lib/sibra/state/reservation.py]: You can't simply sort the indexes, as they wrap. The code doesn't actually care about the numeric value of the indexes at all, so long as you don't try to reuse an active one.

**Pull request merged**